### ISSUE TITLE
Consider showing per-node evaluation time

### ISSUE THREAD

**Issue by juliusv** [promlabs/promlens-public#2]
People are interested in using PromLens to find/debug/prevent costly queries. Currently PromLens shows the number of time series resulting from each sub-expression, which is already a good proxy for evaluation cost (node with many results => probably an expensive node). Consider also showing the per-node evaluation time next to each node, to make the cost even more explicit.

**Comment by MacroPower** [comment]
I think this could be helpful even if results aren't necessarily consistent when evaluating an expression multiple times.

Also, I think it would make sense for any node to show its evaluation time independent of any other factors, if that makes sense. So if I wrap my entire query in `( )`, those parens would have an evaluation time approximately equal to the query inside, versus being marked `0ms`. Also I think that the sum of all nodes doesn't necessarily need to exactly match the time it takes to evaluate the entire query.

I think something more "messy" like this would still add a lot of value:
```
... / ... - 100ms   // the sum of all child eval times do not necessarily need to equal 100ms
    ├─── sum(...) - 60ms
    │    └─── rate(...) - 55ms  // a parent does not necessarily need to be more expensive than a child
    │         └─── example_summary_sum[5m] - 61ms
    └─── sum(...) - 45ms
         └─── rate(...) - 35ms
              └─── example_summary_count[5m] - 30ms
```

**Comment by juliusv** [comment]
@MacroPower That's what I was thinking as well. What do you think of something like this (will be part of the next PromLens release)?

![query_timings](https://user-images.githubusercontent.com/538008/92139123-97fbc080-ee0f-11ea-896f-b0d7972b1ff2.png)

Overall I want to be mindful of not overloading the UI with too much information per node, but this is probably still fine. And options for the future could be to introduce a settings menu (using browser local storage) to configure what info to show, among other settings, or to condense the stats info a bit more and only show more details on click or hover.

**Comment by MacroPower** [comment]
@juliusv That looks great! I think that having those values will be incredibly useful.

> Overall I want to be mindful of not overloading the UI with too much information per node

Completely understand this. That said, I think even without eval times there can be information overload just due to the nature of how you're currently displaying the distinct values per label name. It really depends on the metric, but if for instance I group left on an info metric that has a lot of labels (or just query a metric that has a lot of labels), it can kind of just result in the counts overflowing to the right which isn't ideal for obvious reasons.

![image](https://user-images.githubusercontent.com/5648814/92194497-d35eb500-ee38-11ea-8bb6-5644bd625714.png)

That said I am definitely not a UI/UX person, so yeah, I have no idea if there's a better way to display any of that. Showing details on click/hover seems like it would make sense.

**Comment by juliusv** [comment]
@MacroPower Yeah, for sure, that kind of overflow gets common as soon as anything with containers/k8s gets involved. At the moment my thinking is still that it's valuable enough to always display the labels in detail (actually it's already limited to 10, with more shown on hover). But I'm starting to think more and more that being able to customize this could be a good idea.

**Comment by juliusv** [comment]
@MacroPower I worked on adding a global settings menu for this today:

![settings_menu](https://user-images.githubusercontent.com/538008/92241981-69d5b980-eebf-11ea-86ab-131e037d9c01.gif)

I have more settings I would like to add in the future, so this is just a start.

Settings are stored in browser local storage.

**Comment by juliusv** [comment]
This has been added and released in PromLens 0.9.0 just now (/cc @MacroPower).

Currently it is showing the total end-to-end query time as seen by the browser. In the future, we could introduce more nuanced stats, showing Prometheus's own idea of the internal total evaluation time, and different evaluation phases, as reported by Prometheus's query stats.
